python/paddle/fluid/executor.py

```diff
@@ -253,6 +258,95 @@ def _as_lodtensor(data, place):
     return tensor
 
 
+class _ProgramCompiler(object):
+    def __init__(self, program, place):
+        self._program = program
+        self._place = place
+        self._compiled = False
+        self._is_data_parallel = False
+
+    def _with_data_parallel(self,
+                            loss_name=None,
+                            build_strategy=None,
+                            exec_strategy=None):
+        self._is_data_parallel = True
+        self._places = []
+        self._build_strategy = build_strategy
+        self._exec_strategy = exec_strategy
+        self._loss_name = loss_name
+        self._local_scopes = []
+
+        if self._exec_strategy is None:
+            self._exec_strategy = ExecutionStrategy()
+        if self._build_strategy is None:
+            self._build_strategy = BuildStrategy()
+
+        self._exec_strategy.use_cuda = isinstance(self._place, core.CUDAPlace)
+        if self._exec_strategy.use_cuda:
+            gpus_env = os.getenv("FLAGS_selected_gpus")
+            if gpus_env:
+                gpus = [int(s) for s in gpus_env.split(",")]
+            else:
+                gpus = [
+                    i for i in six.moves.range(core.get_cuda_device_count())
+                ]
+            self._places = [core.CUDAPlace(i) for i in gpus]
+        else:
+            cpu_num = int(
+                os.environ.get('CPU_NUM', multiprocessing.cpu_count()))
+            self._places = [core.CPUPlace() for _ in six.moves.range(cpu_num)]
+        assert self._places, "no place for execution"
+
+        if self._exec_strategy.num_threads == 0:
+            if self._exec_strategy.use_cuda:
+                # Experiments on se-resnext shows that too many threads hurt
+                # performance. Worth tunning for other models in the future.
+                self._exec_strategy.num_threads = len(self._places) * 4
+            else:
+                cpu_num = int(
+                    os.environ.get('CPU_NUM', multiprocessing.cpu_count()))
+                self._exec_strategy.num_threads = cpu_num * 2
+
+        trainers_endpoints = self._program._trainers_endpoints
+        if self._build_strategy.num_trainers > 1 and trainers_endpoints:
+            assert self._build_strategy.num_trainers == len(
+                trainers_endpoints), "num_trainers == len(end_points)"
```

Comment:
The code here indicates the num_trainers should be equal to trainers_endpoints; otherwise assertion will fail. So, to make the condition consistent, should change to "self._build_strategy.num_trainers > 0" since it should align the semantics on whether a list is non empty.

Reply:
@typhoonzero WDYT?

Reply:
Well, when `self._build_strategy.num_trainers` is equal to 1, which means local training, we should not check `trainers_endpoints` though. by the way, `trainers_endpoints` is for sparse GPU gather only.

Reply:
Got it. Probably adding some comments to explain the assumptions.